tokio-executor/src/task/harness.rs

```diff
@@ -0,0 +1,592 @@
+use crate::loom::alloc::Track;
+use crate::loom::cell::CausalCheck;
+use crate::task::waker::waker_ref;
+use crate::task::{raw, Error, Header, Schedule, Snapshot, Task, Trailer};
+
+use std::future::Future;
+use std::marker::PhantomData;
+use std::mem::{ManuallyDrop, MaybeUninit};
+use std::pin::Pin;
+use std::ptr::{self, NonNull};
+use std::task::{Context, Poll, Waker};
+
+/// Typed raw task handle
+pub(super) struct Harness<T, S: 'static> {
+    header: NonNull<Header<S>>,
+    trailer: NonNull<Trailer>,
+    future_or_output: *mut (),
+    _p: PhantomData<(T)>,
+}
+
+impl<T, S> Harness<T, S>
+where
+    T: Future,
+    S: 'static,
+{
+    pub(super) fn from_ptr(
+        header: NonNull<Header<S>>,
+        trailer: NonNull<Trailer>,
+        future_or_output: *mut (),
+    ) -> Harness<T, S> {
+        Harness {
+            header,
+            trailer,
+            future_or_output,
+            _p: PhantomData,
+        }
+    }
+
+    fn header(&self) -> &Header<S> {
+        unsafe { self.header.as_ref() }
+    }
+
+    fn trailer(&self) -> &Trailer {
+        unsafe { self.trailer.as_ref() }
+    }
+}
+
+impl<T, S> Harness<T, S>
+where
+    T: Future,
+    S: Schedule,
+{
+    /// Poll the inner future.
+    ///
+    /// All necessary state checks and transitions are performed.
+    ///
+    /// Panics raised while polling the future are handled.
+    ///
+    /// Returns `true` if the task needs to be scheduled again
+    pub(super) fn poll(self, executor: *const S) -> bool {
```

Comment:
Nit: AFAICT, `executor` is assumed not to be null here. Should it be `NonNull<S>` instead?